lib/Ast.ml

```diff
@@ -1304,6 +1322,18 @@ end = struct
     let dump {ctx; ast= cl} = dump ctx (Cl cl) in
     assert_no_raise ~f:check_cl ~dump xcl
 
+  let check_comprehension
```

Comment:
1. Adding an `| Eexp of Extensions.Expression.t` case?  That'll help in some places, but we still need all the regular machinery for when we don't have an `Ast.t`; I think that might end up confusing more than clarifying, since we'll get the invariant that the `Exp` case is the only expression we *don't* have to check.
2. `fun _ -> true` -- yes, this is wrong, it should be `fun _ -> false`.  (Is this pattern an expression child?  No, it is not.)  I could also make it a true type (`check_comprehension (Pattern pat)` vs. `check_comprehension (Expression exp)`); that would probably be clearest.  It can still be one function, though; the iteration structure is the same, I just have to get my identities right :-)
3. However, I contend the intended semantics of this function are correct: in `[x,y for x = 1 to 10 when x > 3 for y in create ()]`, the following are the direct children of the comprehension:
    * `x,y` (expression, body)
    * `x` (pattern, first iterator)
    * `1` (expression, first iterator)
    * `10` (expression, first iterator)
    * `x > 3` (expression, conditional)
    * `y` (pattern, second iterator)
    * `create ()` (expression, second iterator)
In particular, I think these have to be direct children because there's nothing else for them to be children of.
4. I suppose this could be what you mean by extending `Ast.t`: `| Citer of comprehension_iterator`.  This is probably doable, but I don't think it buys us anything; the fact that they're a child of a comprehension node is enough information to style them correctly, so it would be a lot of work for, as far as I can tell, no gain.

Reply:
Please check out 569f51c and let me know if you're happier!

Reply:
Yes, very happy. Thanks! (And I fully agree about not doing the more-accurate child detection.)